### PR TITLE
Implement DemoteVolume RPC calls.

### PR DESCRIPTION
**Pull request by Madhu-1**
Added functionality to get the parameters and secrets from the volume replication class and call the RPC DemoteVolume.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

### CODE REVIEW

File: controllers/util.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+func (r *VolumeReplicationReconciler) getSecret(name, namespace string) (map[string]string, error) {
+	// Fetch secret
+	namespacedName := types.NamespacedName{Name: name, Namespace: namespace}
+	secret := &corev1.Secret{}
+	err := r.Client.Get(context.TODO(), namespacedName, secret)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Error(err, "Secret not found", "Secret Name", name, "Secret Namespace", namespace)
+		}
+		r.Log.Error(err, "unknown error", "Secret Name", name, "Secret Namespace", namespace)
```
Comment (umangachapagain) [inline controllers/util.go:76]:
```suggestion
		r.Log.Error(err, "error getting secret", "Secret Name", name, "Secret Namespace", namespace)
```

File: controllers/volumereplication_controller.go [hunk 2]
```diff
@@ -84,6 +87,23 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		return ctrl.Result{}, nil
 	}
 
+	// validate the prefix keys in volume replication class
+	parameters, err := removePrefixedParameters(vrcObj.Spec.Parameters)
+	if err != nil {
+		r.Log.Error(err, "failed to remove prefix parameters", "volumeReplicationClass", instance.Spec.VolumeReplicationClass)
```
Comment (umangachapagain) [inline controllers/volumereplication_controller.go:93]: Needs to be rephrased as we are not actually trying to remove prefixed parameters.

**Review by umangachapagain — CHANGES_REQUESTED** [review]

**Review by umangachapagain — COMMENTED** [review]
Are we sure we'll never need context? Will we never do context.Cancel() any task (maybe resync or making volume secondary)

File: controllers/util.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// // Replication Parameters prefixed with replicationParameterPrefix are not passed through
```
Comment (sp98) [inline controllers/util.go]:
```suggestion
	// Replication Parameters prefixed with replicationParameterPrefix are not passed through
```

File: controllers/util.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+				return map[string]string{}, fmt.Errorf("found unknown parameter key \"%q\" with reserved prefix %s", k, replicationParameterPrefix)
+			}
+		} else {
+			newParam[k] = v
+		}
+	}
+	return newParam, nil
+}
+
+func (r *VolumeReplicationReconciler) getSecret(name, namespace string) (map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply controllers/util.go:67]: Not now. let's keep it as a method to VRR if required we can make it generic.

File: controllers/util.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+			// keep adding know prefixes to this list.
+			default:
+				return map[string]string{}, fmt.Errorf("found unknown parameter key \"%q\" with reserved prefix %s", k, replicationParameterPrefix)
+			}
+		} else {
+			newParam[k] = v
+		}
+	}
+	return newParam, nil
+}
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply controllers/util.go:65]: done

**Comment by Madhu-1** [comment]
> Are we sure we'll never need context? Will we never do context.Cancel() any task (maybe resync or making volume secondary)

we are already canceling the context internally, I just removed the ctx which is never used.

File: controllers/secret.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (r *VolumeReplicationReconciler) getSecret(name, namespace string) (map[string]string, error) { // Fetch secret
```
Comment (sp98) [inline controllers/secret.go:27]: comment location `// Fetch secret`

File: controllers/parameters.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+package controllers
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+const (
+	// // Replication Parameters prefixed with replicationParameterPrefix are not passed through
```
Comment (sp98) [inline controllers/parameters.go:26]:
```suggestion
	// Replication Parameters prefixed with replicationParameterPrefix are not passed through
```

**Review by sp98 — CHANGES_REQUESTED** [review]
some nit

File: controllers/parameters.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+				if v == "" {
+					return errors.New("secret name cannot be empty")
+				}
+			case prefixedReplicationSecretNamespaceKey:
+				if v == "" {
+					return errors.New("secret namespace cannot be empty")
+				}
+			// keep adding know prefixes to this list.
+			default:
+				return fmt.Errorf("found unknown parameter key \"%q\" with reserved prefix %s", k, replicationParameterPrefix)
```
Comment (sp98) [inline controllers/parameters.go:64]: %q should add quotes. So we can remove the escape character.

File: controllers/parameters.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+			switch k {
+			case prefixedReplicationSecretNameKey:
+				if v == "" {
+					return errors.New("secret name cannot be empty")
+				}
+			case prefixedReplicationSecretNamespaceKey:
+				if v == "" {
+					return errors.New("secret namespace cannot be empty")
+				}
+			// keep adding know prefixes to this list.
```
Comment (sp98) [inline controllers/parameters.go:62]:
```suggestion
			// keep adding known prefixes to this list.
```

File: controllers/volumereplication_controller.go [hunk 10]
```diff
@@ -84,6 +87,25 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		return ctrl.Result{}, nil
 	}
 
+	err = validatePrefixedParameters(vrcObj.Spec.Parameters)
+	if err != nil {
+		r.Log.Error(err, "failed to validate prefix parameters", "volumeReplicationClass", instance.Spec.VolumeReplicationClass)
+		return ctrl.Result{}, err
+	}
+	// remove the prefix keys in volume replication class  parameters
```
Comment (sp98) [inline controllers/volumereplication_controller.go:95]:
```suggestion
	// remove the prefix keys in volume replication class parameters
```

**Review by sp98 — CHANGES_REQUESTED** [review]
A few more nits. Looks good otherwise.

File: controllers/parameters.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+// replicationclass.
+func removePrefixedParameters(param map[string]string) map[string]string {
+	newParam := map[string]string{}
+	for k, v := range param {
+		if !strings.HasPrefix(k, replicationParameterPrefix) {
+			newParam[k] = v
+		}
+	}
+	return newParam
+}
```
Comment (umangachapagain) [inline controllers/parameters.go:46]:
```suggestion
func filterPrefixedParameters(prefix string, param map[string]string) map[string]string {
	newParam := map[string]string{}
	for k, v := range param {
		if !strings.HasPrefix(k, prefix) {
			newParam[k] = v
		}
	}
	return newParam
}
```

**Review by sp98 — APPROVED** [review]

File: controllers/parameters.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+	prefixedReplicationSecretNameKey      = replicationParameterPrefix + "replication-secret-name"      // name key for secret
+	prefixedReplicationSecretNamespaceKey = replicationParameterPrefix + "replication-secret-namespace" // namespace key secret
+)
+
+// filterPrefixedParameters removes all the reserved keys from the
+// replicationclass which are matching the prefix.
+func filterPrefixedParameters(prefix string, param map[string]string) map[string]string {
+	newParam := map[string]string{}
+	for k, v := range param {
+		if !strings.HasPrefix(k, replicationParameterPrefix) {
```
Comment (umangachapagain) [inline controllers/parameters.go:41]:
```suggestion
		if !strings.HasPrefix(k, prefix) {
```

**Review by umangachapagain — CHANGES_REQUESTED** [review]

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**